core-bundle/src/Routing/PreviewUrlGenerator.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Routing;
+
+use Contao\ArticleModel;
+use Contao\CoreBundle\Event\ContaoCoreEvents;
+use Contao\CoreBundle\Event\PreviewUrlCreateEvent;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+class PreviewUrlGenerator
+{
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var ContaoFramework
+     */
+    private $framework;
+
+    public function __construct(EventDispatcherInterface $eventDispatcher, RouterInterface $router, RequestStack $requestStack, ContaoFramework $framework)
+    {
+        $this->eventDispatcher = $eventDispatcher;
+        $this->router = $router;
+        $this->requestStack = $requestStack;
+        $this->framework = $framework;
+    }
+
+    public function getPreviewUrl(): string
+    {
+        if (!$request = $this->requestStack->getCurrentRequest()) {
+            throw new \RuntimeException('The request stack did not contain a request');
```

Comment:
Is that really an issue here? Maybe we should allow to return `null` and in that case, the URL is just not added to the header menu?

Reply:
This is the default exception that we throw everywhere where we retrieve the current request from the request stack. It should never be thrown in real life.

Reply:
Ok.